src/jit/morph.cpp

```diff
@@ -17987,422 +17971,728 @@ GenTree* Compiler::fgMorphImplicitByRefArgs(GenTree* tree, bool isAddr)
+                break;
+
+            case GT_OBJ:
+            case GT_BLK:
+            case GT_IND:
+                assert(TopValue(1).Node() == node);
+                assert(TopValue(0).Node() == node->gtGetOp1());
+
+                if (!TopValue(1).Indir(TopValue(0)))
                 {
-                    axcStack->Push(AXC_IndWide);
+                    // If the address comes from another indirection (e.g. IND(IND(...))
+                    // then we need to escape the location.
+                    EscapeLocation(TopValue(0), node);
                 }
-                else
+
+                PopValue();
+                break;
+
+            case GT_DYN_BLK:
+                assert(TopValue(2).Node() == node);
+                assert(TopValue(1).Node() == node->AsDynBlk()->Addr());
+                assert(TopValue(0).Node() == node->AsDynBlk()->gtDynamicSize);
+
+                // The block size may be the result of an indirection so we need
+                // to escape the location that may be associated with it.
+                EscapeValue(TopValue(0), node);
+
+                if (!TopValue(2).Indir(TopValue(1)))
                 {
-                    axcStack->Push(AXC_Ind);
+                    // If the address comes from another indirection (e.g. DYN_BLK(IND(...))
+                    // then we need to escape the location.
+                    EscapeLocation(TopValue(1), node);
                 }
-                return WALK_CONTINUE;
-            }
 
-        case GT_LCL_FLD:
-        {
-            assert(axc != AXC_Addr);
-            unsigned lclNum = tree->gtLclVarCommon.gtLclNum;
-            if (comp->lvaIsImplicitByRefLocal(lclNum))
-            {
-                // Keep track of the number of appearances of each promoted implicit
-                // byref (here during address-exposed analysis); fgMakeOutgoingStructArgCopy
-                // checks the ref counts for implicit byref params when deciding if it's legal
-                // to elide certain copies of them.
-                LclVarDsc* varDsc = &comp->lvaTable[lclNum];
-                JITDUMP("Incrementing ref count from %d to %d for V%02d in fgMorphStructField\n",
-                        varDsc->lvRefCnt(RCS_EARLY), varDsc->lvRefCnt(RCS_EARLY) + 1, lclNum);
+                PopValue();
+                PopValue();
+                break;
 
-                varDsc->incLvRefCnt(1, RCS_EARLY);
-            }
-            // This recognizes certain forms, and does all the work.  In that case, returns WALK_SKIP_SUBTREES,
-            // else WALK_CONTINUE.  We do the same here.
-            fgWalkResult res = comp->fgMorphLocalField(tree, fgWalkPre);
-            if (res == WALK_SKIP_SUBTREES && tree->OperGet() == GT_LCL_VAR && (axc == AXC_Addr || axc == AXC_AddrWide))
-            {
-                comp->lvaSetVarAddrExposed(lclNum);
-                if (axc == AXC_AddrWide)
+            default:
+                while (TopValue(0).Node() != node)
```

Comment:
Still don't understand:
For sequence of `IND(ADD(4, ADDR(LCL_VAR)))` preorder will generate such stack:
`VALUE(LCL_VAR), VALUE(ADDR), VALUE(4), VALUE(ADD), VALUE(IND)`, is it right?

Then we start post order with `LCL_VAR` and replace `VALUE(LCL_VAR)` with `Location(LCL_VAR)`, 
then we have `ADDR` that replaces `VALUE(ADDR)` with `Address(LCL_VAR)` and pop `Location(LCL_VAR)` from the stack, is it right?

At this moment on the stack we have: `Address(LCL_VAR), VALUE(4), VALUE(ADD), VALUE(IND)`, correct?

Then we come to `4` and go to this default case that checks `while (TopValue(0).Node() != node)`, and it will pop `TopValue(0)` that is `Address(LCL_VAR)` from the stack, is it correct?

Also it will call `EscapeValue(Address(LCL_VAR), `4`); ` that will work fine but is not semantically correct  because `4` is not `//    user - the node that uses the escaped value`, correct?

> To add support for ADD we need to pop its 2 operand values. 

I think we don't have 2 operand values on the stack when we come to ADD because we have poped 'ADDRESS(LCL_VAR)` in the default case that processed `4`.

Reply:
> For sequence of IND(ADD(4, ADDR(LCL_VAR))) preorder will generate such stack:
VALUE(LCL_VAR), VALUE(ADDR), VALUE(4), VALUE(ADD), VALUE(IND), is it right?

Yes. But it seems that you mixed up the post order traversal. When you reach 4 it goes like this:
```
pre 4: push 4 -> VALUE(4), VALUE(ADD), VALUE(IND)
post 4: leaves itself on top of the stack -> VALUE(4), VALUE(ADD), ...
pre ADDR: push ADDR -> VALUE(ADDR), VALUE(4), ...
pre LCL_VAR: push -> LCL_VAR -> VALUE(LCL_VAR), VALUE(ADDR), VALUE(4), ...
post LCL_VAR: leaves itself on top of the stack -> VALUE(LCL_VAR), VALUE(ADDR), VALUE(4), ...
post ADDR: pop LCL_VAR and leaves itself on top of the stack -> VALUE(ADDR), VALUE(4), VALUE(ADD), VALUE(IND)
post ADD: pops ADDR and 4 until it reaches itself and stay on top of the stack -> VALUE(ADD), VALUE(IND)
```
In summary - you do post order when all the children have been processed. That includes post order having ran on children. Post order leaves the node on the stack at the end so that means that at the beginning of post order on top of the stack we have the child node values, one for each node, followed by the post order node value itself.

> I think we don't have 2 operand values on the stack when we come to ADD because we have poped 'ADDRESS(LCL_VAR)in the default case that processed4`.

This doesn't handle ADD now but it does handle DYN_BLK and that one has 3 operands. If the stack was messed up then one of its asserts must have fired until now.

Reply:
I should add that post order is the natural ordering for expression tree evaluation. Normally it's done recursively - EvalAdd - Eval(op1) + Eval(op2) - but that assumes that the code knows how many operands a node has. 

That happens to be somewhat problematic with JIT's IR. You could do something like `for (GenTree* op : node->Operands()) { EscapeValue(PopValue()); }` but that's slower and ultimately redundant since the tree visitor knows the operands. So I just made it so that pre order leaves breadcrumbs on the evaluation stack so post order knows how many nodes to pop.

Reply:
Yes, you are right, thank you!